onboarding/mlsys/1-gradients.md

```diff
@@ -0,0 +1,493 @@
+
+```python
+# Two workers, each with two batches of two data points each.
+shards = [
+    [
+        [(1,1),(2,2)],     # first shard, first batch
+        [(3,3), (4,4)],    # first shard, second batch
+    ],
+    [
+        [(1,0),(2,1)],     # second shard, first batch
+        [(3,1), (4,6.25)], # second shard, second batch
+    ]
+]
+
+mfinal = train_loop_distributed(0, 0.001, shards, iterations=4)
+assert round(mfinal, 5) == 0.05893, mfinal
+```
+
+Notice that you can achieve the same exact results with the
+`train_loop_batches` you wrote previously:
+
+```python
+dataset = [
+    [(1,1),(2,2), (1,0),(2,1)],
+    [(3,3), (4,4), (3,1), (4,6.25)],
+]
+mfinal = train_loop_batches(m=0, lr=0.001, dataset=dataset, iterations=4)
+assert round(mfinal, 5) == 0.05893, mfinal
+```
+
+Explain why you see the same training result in both cases.
+
+## Data-Parallel Communication Overhead
+
+Because data-parallel dtrain communicates gradients after every single step, it
+is most effective with models that spend a relatively long time calculating
+each gradient.  If the gradient calculation step is relatively fast compared to
+the communication time between workers, dtrain can cause significant slowdowns
+to training.  This is why distributed MNIST models don't really make any sense;
+the model is so small and fast it spends most of its time communicating rather
+than training.
+
+There are specific model features which can aggravate this problem, such as the
+presence of a SyncBatchNormalization layer in the model.
+
+Users will occasionally not understand how these communication overheads affect
+training, and will ask questions like, "I started training my model on 2 GPUs
+and it got slower", so understanding the overhead costs helps you answer those
+sorts of questions.
+
+## Gradient Aggregation
+
+Determined supports "gradient aggregation", where you calculate multiple
+batches of gradients in sequence, combine them with averaging, and then take a
+step based on that "aggregated" gradient.
+
+You should be able to see that the steps involved in gradient aggregation are
+essentially identical to the `train_loop_distributed` function you just
+implemented.  If data-parallel dtrain is a way to simulate larger batch sizes
+across multiple workers, gradient aggregation is a way to simulate larger batch
+sizes across multiple iterations within a single worker.
+
+Unfortunately, gradient aggregation isn't particularly useful by itself.  When
+training on a single worker, applying gradients doesn't take very long, so it
+doesn't make much sense to not apply gradients from a batch as soon as you
+finish calculating that batch.
+
+However, it is commonly applied to data-parallel dtrain, because it directly
+reduces network communications.  If you always aggregate gradients from 4
+batches before applying them to the model, you only need to communicate
+gradients between workers 1/4th of the time, significantly reducing your
+communication overhead.
```

Comment:
I would recommend adding some kind of summary at the end to wrap it up, similar to the paragraph we had at the beginning: 

"Afterward, you will understand basic concepts like what a gradient is and how
gradient descent works to train models.  You will also understand how many
features in Determined, like batch size, gradient aggregation, and
data-parallel distributed training, really just boil down to different ways of
averaging gradients together during gradient descent."

Mostly a nit, I just felt it ended abruptly

Reply:
great idea